src/libcompiler_builtins/lib.rs

```diff
@@ -17,3 +17,644 @@
+        fn to_bytes(self) -> u64 { unsafe { ::core::mem::transmute(self) } }
+        fn get_exponent(self) -> i32 {
+            (((self.to_bytes() & Self::EXP_MASK) >> Self::MANTISSA_BITS) as i32) - Self::MAX_EXP
+        }
+    }
+
+    macro_rules! float_as_unsigned {
+        ($from: expr, $fromty: ty, $outty: ty) => { {
+            use core::num::Float;
+            let repr = $from.to_bytes();
+            let sign = $from.signum();
+            let exponent = $from.get_exponent();
+            let mantissa_fraction = repr & <$fromty as FloatStuff>::MANTISSA_MASK;
+            let mantissa = mantissa_fraction | <$fromty as FloatStuff>::MANTISSA_LEAD_BIT;
+            if sign == -1.0 || exponent < 0 { return 0; }
+            if exponent > ::core::mem::size_of::<$outty>() as i32 * 8 {
+                return !0;
+            }
+            if exponent < (<$fromty as FloatStuff>::MANTISSA_BITS) as i32 {
+                mantissa as $outty >> (<$fromty as FloatStuff>::MANTISSA_BITS as i32 - exponent)
+            } else {
+                mantissa as $outty << (exponent - <$fromty as FloatStuff>::MANTISSA_BITS as i32)
+            }
+        } }
+    }
+
+    #[export_name="__fixunsdfti"]
+    pub extern fn f64_as_u128(a: f64) -> u128_ {
+        float_as_unsigned!(a, f64, u128_)
+    }
+
+    #[export_name="__fixunssfti"]
+    pub extern fn f32_as_u128(a: f32) -> u128_ {
+        float_as_unsigned!(a, f32, u128_)
+    }
+
+    macro_rules! float_as_signed {
+        ($from: expr, $fromty: ty, $outty: ty) => {{
+            use core::num::Float;
+            let repr = $from.to_bytes();
+            let sign = $from.signum();
+            let exponent = $from.get_exponent();
+            let mantissa_fraction = repr & <$fromty as FloatStuff>::MANTISSA_MASK;
+            let mantissa = mantissa_fraction | <$fromty as FloatStuff>::MANTISSA_LEAD_BIT;
+
+            if exponent < 0 { return 0; }
+            if exponent > ::core::mem::size_of::<$outty>() as i32 * 8 {
+                return if sign > 0.0 { <$outty>::max_value() } else { <$outty>::min_value() };
+            }
+            let r = if exponent < (<$fromty as FloatStuff>::MANTISSA_BITS) as i32 {
+                mantissa as $outty >> (<$fromty as FloatStuff>::MANTISSA_BITS as i32 - exponent)
+            } else {
+                mantissa as $outty << (exponent - <$fromty as FloatStuff>::MANTISSA_BITS as i32)
+            };
+            if sign >= 0.0 { r } else { -r }
+        }}
+    }
+
+    #[export_name="__fixdfti"]
+    pub extern fn f64_as_i128(a: f64) -> i128_ {
+        float_as_signed!(a, f64, i128_)
+    }
+
+    #[export_name="__fixsfti"]
+    pub extern fn f32_as_i128(a: f32) -> i128_ {
+        float_as_signed!(a, f32, i128_)
+    }
+
+    #[export_name="__floattidf"]
+    pub extern fn i128_as_f64(a: i128_) -> f64 {
+        match a.signum() {
+            1 => u128_as_f64(a.abs() as u128_),
+            0 => 0.0,
+            -1 => -u128_as_f64(a.abs() as u128_),
```

Comment:
It doesn't really work well for `__muloti4` as you have to negate the value, but I've changed it everywhere else. For `__muloti4` its no big deal either way as we already check it manually beforehand. To make sure that there is really no panic (LLVM may remove the call but it may not, which then will lead to a linker error later on) I've made a non panicking `iabs` method